src/mesh/p4est_mesh.jl

```diff
@@ -397,3 +399,239 @@ end
+  @timed timer() "refine" p4est_refine(mesh.p4est, false, refine_fn_c, init_fn_c)
+
+  @timed timer() "rebalance" p4est_balance(mesh.p4est, P4EST_CONNECT_FACE, init_fn_c)
+  # Due to a bug in p4est, the forest needs to be rebalanced twice sometimes
+  # See https://github.com/cburstedde/p4est/issues/112
+  @timed timer() "rebalance" p4est_balance(mesh.p4est, P4EST_CONNECT_FACE, init_fn_c)
+
+  return collect_changed_cells(mesh, original_n_cells)
+end
+
+
+function coarsen_fn(p4est, which_tree, quadrants_ptr)
+  quadrants = unsafe_wrap(Array, quadrants_ptr, 4)
+
+  # Controller value has been copied to the quadrant's user data storage before.
+  # Unpack quadrant's user data ([global quad ID, controller_value])
+  if all(i -> unsafe_load(Ptr{Int}(quadrants[i].p.user_data), 2) < 0, eachindex(quadrants))
+    # return true (coarsen)
+    return Cint(1)
+  else
+    # return false (don't coarsen)
+    return Cint(0)
+  end
+end
+
+# Coarsen marked cells if the forest will stay balanced.
+# Return a list of all cells that have been coarsened.
+function coarsen!(mesh::P4estMesh)
+  # Copy original element IDs to quad user data storage
+  original_n_cells = ncells(mesh)
+  save_original_ids(mesh)
+
+  # Coarsen marked cells
+  coarsen_fn_c = @cfunction(coarsen_fn, Cint,
+    (Ptr{p4est_t}, Ptr{p4est_topidx_t}, Ptr{Ptr{p4est_quadrant_t}}))
+  init_fn_c = @cfunction(init_fn, Cvoid,
+    (Ptr{p4est_t}, Ptr{p4est_topidx_t}, Ptr{p4est_quadrant_t}))
+
+  @timed timer() "coarsen!" p4est_coarsen(mesh.p4est, false, coarsen_fn_c, init_fn_c)
+
+  new_cells = collect_new_cells(mesh)
+  coarsened_cells_vec = collect_changed_cells(mesh, original_n_cells)
+  # 2^NDIMS changed cells should have been coarsened to one new cell.
+  # This matrix will store the IDs of all cells that have been coarsened to cell new_cells[i]
+  # in the i-th column.
+  coarsened_cells = reshape(coarsened_cells_vec, 2^ndims(mesh), length(new_cells))
+
+  # Save new original IDs to find out what changed after balancing
+  intermediate_n_cells = ncells(mesh)
+  save_original_ids(mesh)
+
+  @timed timer() "rebalance" p4est_balance(mesh.p4est, P4EST_CONNECT_FACE, init_fn_c)
+  # Due to a bug in p4est, the forest needs to be rebalanced twice sometimes
+  # See https://github.com/cburstedde/p4est/issues/112
+  @timed timer() "rebalance" p4est_balance(mesh.p4est, P4EST_CONNECT_FACE, init_fn_c)
+
+  refined_cells = collect_changed_cells(mesh, intermediate_n_cells)
+
+  # Some cells may have been coarsened even though they unbalanced the forest.
+  # These cells have now been refined again by p4est_balance.
+  # refined_cells contains the intermediate IDs (ID of coarse cell
+  # between coarsening and balancing) of these cells.
+  # Find original ID of each cell that has been coarsened and then refined again.
+  for refined_cell in refined_cells
+    # i-th cell of the ones that have been created by coarsening has been refined again
+    i = findfirst(isequal(refined_cell), new_cells)
```

Comment:
```suggestion
    i = findfirst(==(refined_cell), new_cells)
```
We usually don't use `isequal` but `==` in the rest of the code